src/Uno.Toolkit.Skia.WinUI/Controls/Shadows/ShadowContainer.Paint.cs

```diff
@@ -287,96 +182,122 @@ private static void DrawDropShadow(SKCanvas canvas, SKPaint paint, SKShadow shad
 	}
 
-	private bool TryGetContentBackground(out Brush? background)
+	/// <summary>
+	/// Record of <see cref="Shadow"/> properties at one point in time.
+	/// </summary>
+	private readonly record struct ShadowInfo(bool Inset, float OffsetX, float OffsetY, float BlurRadius, float Spread, Color Color, float Opacity)
 	{
-		if (_currentContent == null)
-		{
-			background = null;
-			return false;
-		}
+		public static ShadowInfo Snapshot(Shadow x) => new(
+			x.IsInner,
+			(float)x.OffsetX, (float)x.OffsetY,
+			(float)x.BlurRadius, (float)x.Spread,
+			x.Color,
+			(float)x.Opacity
+		);
 
-		background = _currentContent switch
-		{
-			Control control => control.Background,
-			Panel panel => panel.Background,
-			Border border => border.Background,
-			_ => null,
-		};
+		public static ShadowInfo[] Snapshot(IEnumerable<Shadow> shadows) => shadows.Safe().Select(Snapshot).ToArray();
 
-		return background != null;
-	}
+		public SKColor GetSKColor() => new SKColor(Color.R, Color.G, Color.B, (byte)(Color.A * Opacity));
 
-	private bool TrySetContentBackground(SolidColorBrush background)
-	{
-		switch (_currentContent)
+		public float GetBlurSigma(float pixelRatio)
 		{
-			case Control control:
-				control.Background = background;
-				break;
-			case Panel panel:
-				panel.Background = background;
-				break;
-			case Border border:
-				border.Background = background;
-				break;
-			default:
-				return false;
+			var radius = BlurRadius * pixelRatio;
+			// Blur sigma conversion taken from flutter source code
+			var sigma = radius > 0 ? radius * 0.57735f + 0.5f : 0f;
+
+			return sigma;
 		}
 
-		return true;
+		public string ToKey() => string.Join(",", Inset, OffsetX, OffsetY, BlurRadius, Spread, Color, Opacity);
 	}
 
-	private static SKColor ToSkiaColor(Color windowsUiColor)
+	/// <summary>
+	/// Used in comparison to determine if the shadow needs to be repainted.
+	/// </summary>
+	private sealed record ShadowPaintState(
+		float ContentWidth, float ContentHeight,
+		Color? Background,
+		CornerRadius CornerRadius,
+		float PixelRatio,
+		ShadowInfo[] ShadowInfos)
 	{
-		return new SKColor(windowsUiColor.R, windowsUiColor.G, windowsUiColor.B, windowsUiColor.A);
+		public IEnumerable<ShadowInfo> GetDropShadows() => ShadowInfos.Safe().Where(x => !x.Inset);
+		public IEnumerable<ShadowInfo> GetInsetShadows() => ShadowInfos.Safe().Where(x => x.Inset);
+
+		// GetHashCode+Equals are needed here for ShadowInfos' sequential equality
+
+		public override int GetHashCode()
+		{
+			var hash = 1214348419;
+
+			hash = hash * -1521134295 + ContentWidth.GetHashCode();
+			hash = hash * -1521134295 + ContentHeight.GetHashCode();
+			hash = hash * -1521134295 + PixelRatio.GetHashCode();
+			hash = hash * -1521134295 + Background.GetHashCode();
+			hash = hash * -1521134295 + CornerRadius.GetHashCode();
+			foreach (var item in ShadowInfos)
+			{
+				hash = hash * -1521134295 + item.GetHashCode();
+			}
+
+			return hash;
```

Comment:
How about using the built-in `HashCode` struct from the [BCL](https://learn.microsoft.com/en-us/dotnet/api/system.hashcode?view=net-7.0&viewFallbackFrom=dotnet-plat-ext-3.1#moniker-applies-to)? If I'm not wrong this was added on netstandard 2.1
```suggestion
			var hashCode = new HashCode();

			hashCode.Add(ContentWidth);
			hashCode.Add(ContentHeight);
			hashCode.Add(PixelRatio);
			hashCode.Add(Background);
			hashCode.Add(CornerRadius);

			foreach (var item in ShadowInfos)
			{
				hashCode.Add(item);
			}

			return hashCode.ToHashCode();
```